### ISSUE TITLE
Legendary error on Epic Games login and cannot login (snap)

### ISSUE THREAD

**Issue by Skythrew** [Heroic-Games-Launcher/HeroicGamesLauncher#2573]
### Describe the bug

I installed Heroic Games Launcher's snap on Ubuntu for testing purposes today instead of getting the deb package and it seems that Epic Games login is not functional because I get a "Permission denied" which is probably caused by snapd's restrictions.

### Add logs

```text
(12:44:40) INFO:    [Connection]:       Connectivity: check-online
(12:44:40) INFO:    [Connection]:       Pinging external endpoints
(12:44:40) INFO:    [Legendary]:        Running command: /snap/heroic/9/resources/app.asar.unpacked/build/bin/linux/legendary --version
(12:44:40) INFO:    [Legendary]:        Legendary location: /snap/heroic/9/resources/app.asar.unpacked/build/bin/linux/legendary
(12:44:40) INFO:    [Gog]:              GOGDL location: /snap/heroic/9/resources/app.asar.unpacked/build/bin/linux/gogdl
(12:44:40) INFO:    [Gog]:              GOGDL location: /snap/heroic/9/resources/app.asar.unpacked/build/bin/linux/gogdl
(12:44:40) WARNING: [Backend]:          Protocol already registered.
(12:44:40) INFO:    [Connection]:       Connectivity: online
(12:44:40) INFO:    [Backend]:          User Not Found, removing it from Store
(12:44:40) INFO:    [Gog]:              Running command: /snap/heroic/9/resources/app.asar.unpacked/build/bin/linux/gogdl --auth-config-path /home/mael/snap/heroic/9/.config/heroic/gog_store/auth.json --version
(12:44:41) INFO:    [Backend]:          

Heroic Version: 2.6.3 Trafalgar Law
Legendary Version:  0.20.32 Dark Energy (hotfix #6)
GOGdl Version: 

Electron Version: 23.1.4
Chrome Version: 110.0.5481.192
NodeJS Version: 18.12.1

OS: Ubuntu Core KERNEL: 5.19.0-38-generic ARCH: x64
CPU: AMD Ryzen 7 5700G with Radeon Graphics @3.77 GOVERNOR: schedutil
RAM: Total: 30.75 GiB Available: 26.31 GiB
GRAPHICS: 
PROTOCOL: wayland

(12:44:41) INFO:    [Backend]:          AreWeAntiCheatYet data downloaded
(12:44:41) INFO:    [Backend]:          Loading Screen Ready
(12:44:41) INFO:    [Backend]:          Frontend Ready
(12:44:41) INFO:    [Backend]:          Checking for new Heroic Updates
(12:44:41) INFO:    [Backend]:          Checking for current version changelog
(12:44:41) ERROR:   [Backend]:          Error when checking for current Heroic changelog
(12:44:43) INFO:    [Winetricks]:       Downloading Winetricks
(12:44:43) INFO:    [DXVKInstaller]:    Updating vkd3d to: vkd3d-proton-v2.8
(12:44:43) WARNING: [DXVKInstaller]:    Error when downloading vkd3d Error: Command failed: curl -L https://github.com/Heroic-Games-Launcher/vkd3d-proton/releases/download/v2.8/vkd3d-proton-v2.8.tar.xz -o '/home/mael/snap/heroic/9/.config/heroic/tools/vkd3d/vkd3d-proton-v2.8.tar.xz' --create-dirs
/bin/sh: 1: curl: not found

    at ChildProcess.exithandler (node:child_process:422:12)
    at ChildProcess.emit (node:events:513:28)
    at maybeClose (node:internal/child_process:1091:16)
    at Socket.<anonymous> (node:internal/child_process:449:11)
    at Socket.emit (node:events:513:28)
    at Pipe.<anonymous> (node:net:313:12)
(12:44:43) INFO:    [DXVKInstaller]:    Updating dxvk to: dxvk-2.1
(12:44:43) WARNING: [DXVKInstaller]:    Error when downloading dxvk Error: Command failed: curl -L https://github.com/doitsujin/dxvk/releases/download/v2.1/dxvk-2.1.tar.gz -o '/home/mael/snap/heroic/9/.config/heroic/tools/dxvk/dxvk-2.1.tar.gz' --create-dirs
/bin/sh: 1: curl: not found

    at ChildProcess.exithandler (node:child_process:422:12)
    at ChildProcess.emit (node:events:513:28)
    at maybeClose (node:internal/child_process:1091:16)
    at Socket.<anonymous> (node:internal/child_process:449:11)
    at Socket.emit (node:events:513:28)
    at Pipe.<anonymous> (node:net:313:12)
(12:44:46) INFO:    [Backend]:          Starting the Download Queue
(12:45:28) INFO:    [Backend]:          Heroic: Setting discordRPC to true
(12:46:14) INFO:    [Legendary]:        Logging in: /snap/heroic/9/resources/app.asar.unpacked/build/bin/linux/legendary auth --code <redacted>
(12:46:15) INFO:    [Frontend]:         Refreshing Library
(12:46:15) INFO:    [Legendary]:        Refreshing library...
(12:46:15) INFO:    [Frontend]:         No cache found, getting data from legendary...
(12:46:15) INFO:    [Legendary]:        Refreshing library...
(12:47:12) INFO:    [Legendary]:        Logging in: /snap/heroic/9/resources/app.asar.unpacked/build/bin/linux/legendary auth --code <redacted>
(12:47:12) INFO:    [Frontend]:         Refreshing Library
(12:47:12) INFO:    [Legendary]:        Refreshing library...
(12:47:12) INFO:    [Frontend]:         No cache found, getting data from legendary...
(12:47:12) INFO:    [Legendary]:        Refreshing library...
```

### Steps to reproduce

1. Go to 'Log in'
2. Click on 'Epic Games Login'
3. Log into your account

### Expected behavior

We should be able to login to our Epic Games account.

### Screenshots

![Capture d’écran du 2023-03-26 12-47-20](https://user-images.githubusercontent.com/76571984/227771018-2b65378d-db8c-450b-994e-5f67468abb8f.png)

### Heroic Version

Latest Stable

### System Information

- OS: Ubuntu 22.10

**Comment by flavioislima** [comment]
Thanks for reporting that, we will look into it 👍

**Comment by Saroumane** [comment]
Same result here (Ubuntu 22.10), Heroic Games Launcher in snap version 2.6.3 seems to be broken, because of a PATH or permission problem to access `curl`.
I also tried (with same result) `snap install --devmode --edge heroic`

**Comment by Illidan1943** [comment]
Same here in Windows 11 in 2.7

**Comment by w-sky** [comment]
Just updated to 2.7.1
Still getting this error when trying to log in, using alternative login method. Snap install on Ubuntu 22.04. The errors are not all identical:

`An error has occurred! Try to Logout and Login on your Epic account.
Traceback (most recent call last):
File "legendary/cli.py", line 3061, in <module>
File "legendary/cli.py", line 2945, in main
File "legendary/cli.py", line 48, in setup_threaded_loggingFile "multiprocessing/context.py", line 103, in Queue
File "multiprocessing/queues.py", line 43, in __init__
File "multiprocessing/context.py", line 68, in Lock
File "multiprocessing/synchronize.py", line 162, in __init__
File "multiprocessing/synchronize.py", line 57, in __init__
PermissionError: [Errno 13] Permission denied
[9810] Failed to execute script 'cli' due to unhandled exception!`

**Comment by guiiamorim** [comment]
I'm having the same issue with version 2.7.1 on Ubuntu 22.04.

```
An error has occurred! Try to Logout and Login on your Epic account.
Traceback (most recent call last):
File "legendary/cli.py", line 3061, in <module>
File "legendary/cli.py", line 2945, in main
File "legendary/cli.py", line 48, in setup_threaded_logging
File "multiprocessing/context.py", line 103, in Queue
File "multiprocessing/queues.py", line 43, in __init__
File "multiprocessing/context.py", line 68, in Lock
File "multiprocessing/synchronize.py", line 162, in __init__
File "multiprocessing/synchronize.py", line 57, in __init__
PermissionError: [Errno 13] Permission denied
[18854] Failed to execute script 'cli' due to unhandled exception!
```

Also, I have another error when trying to download wine/proton versions, but I don't know if it's related to this.

**Comment by arielj** [comment]
@flavioislima the snap support was removed because of too many issues, right? can we close this?

**Comment by flavioislima** [comment]
> @flavioislima the snap support was removed because of too many issues, right? can we close this?

Not really, I just need to reach out the snap team so they can enable some access for heroic snap

**Comment by ivocavalcante** [comment]
Hi people. Where can I find the **snapcraft.yaml** file? Want to dig a little into this problem, but couldn't find it. Apologize in advance if it's in an obvious place...

**Comment by imLinguin** [comment]
We actually set it up using electron builder's config.  
https://github.com/Heroic-Games-Launcher/HeroicGamesLauncher/blob/main/package.json#L132  
https://www.electron.build/configuration/snap.html

**Comment by ivocavalcante** [comment]
Great, that was fast. Let me look into it, see if something sparks.

**Comment by ivocavalcante** [comment]
From quick glance, I can see that, **apparently**, "curl" hasn't been staged into the snap, and could be the cause of OP error. Guess it'll need some tweaking. I'm not familiar with Electron build system, but will give a shot.

**Comment by imLinguin** [comment]
That still leaves the python multiprocessing `Permission denied` error

**Comment by RobertoNovo** [comment]
I was getting this same error using the "Snap Shop" version of HeroicGamesLauncher
I removed that version and downloaded the AppImage from
https://heroicgameslauncher.com/downloads

I run this to give permissions of execution (as chatGPT suggested) in the same folder as the AppImage:
chmod +x Heroic-2.8.0.AppImage

Then
./Heroic-2.8.0.AppImage

This returned me errors missing FUSE so I installed it from
https://github.com/AppImage/AppImageKit/wiki/FUSE

Tried again
./Heroic-2.8.0.AppImage

And it just run the program. I logged in normally using my epicgames account and I can see my games in the library. I haven't gone any further yet.

I hope it will be useful to someone

**Comment by imLinguin** [comment]
@ivocavalcante Would you be able to test this pull request? https://github.com/Heroic-Games-Launcher/HeroicGamesLauncher/pull/2826

I think it solved curl issue. Not sure about Python things.  
I have some trouble setting up snap on my machine, that's why ask for help :)

**Comment by ivocavalcante** [comment]
Sure, will do. Report back soon.

**Comment by ivocavalcante** [comment]
Ok @imLinguin , did some tests.

I could'n generate the package using your branch, got an error regarding "DISABLE_WAYLAND" environment variable (or something like that). Thought it could be related to the old base being used on the Snap - **core18**, which is EOL and supported only in legacy mode. Tried to update to core20 or 22, newer bases, but electron-builder kept complaining that it couldn't recognize the key "base" on packages.json ; had to update builder version to newest release (24.4.0) and it was recognized then.After compiling the package on core22, I noticed that, even though "curl" was listed as a staged packaged on json, it hasn't really been included on the Snap. This, plus the fact that, despite now using core22, electron-builder declared plugs on a version of Gnome content snap related to core18 and also the lack of support for Snap extensions (among other issues) made me give up on this approach.

Tl;DR
I don't believe that electron-builder is up to this task. If you people are serious about having a Snap package - instead of "builder creates it, so why not?", if this was the case - I think that having a explicit snapcraft.yaml recipe is the right approach (either using options two or three of this [link](https://snapcraft.io/blog/three-ways-to-package-your-electron-apps-as-snaps)). If you decide to take this route, I might be able to help.

Obviously, this is My Humble Opinion(tm) and I had zero knowledge on Electron build systems before this, so take my words with a grain of salt.

**Comment by arielj** [comment]
I understand this should be fixed now in 2.10.0 since it includes some Span fixes

**Comment by Valentin-Ronsseray** [comment]
The issue has not been fixed at all :

> OS: Debian GNU/Linux 12 (bookworm) (linux)
> 
> The current system is not a Steam Deck
> We are not running inside a Flatpak container
> 
> Software Versions:
>   **Heroic: 2.10.0 Jewelry Bonney**
>   Legendary: 0.20.33 Undue Alarm
>   gogdl: 0.7.3
>   Nile: 1.0.0 Jonathan Joestar

**Comment by flavioislima** [comment]
> The issue has not been fixed at all :
> 
> > OS: Debian GNU/Linux 12 (bookworm) (linux)
> > The current system is not a Steam Deck
> > We are not running inside a Flatpak container
> > Software Versions:
> > **Heroic: 2.10.0 Jewelry Bonney**
> > Legendary: 0.20.33 Undue Alarm
> > gogdl: 0.7.3
> > Nile: 1.0.0 Jonathan Joestar

We need the logs from when you try to login with Epic.
On Logs get the Legendary/Epic Logs.
